### PR TITLE
Add Jenkins CI to the repo with code refactoring

### PR DESCRIPTION
**Pull request by pydevops**
- code refactoring for Jenkins CI
    - common.sh
    - validate.sh 
    - delete.sh: make sure bq delete the dataset prior to the terraform destroy. 
- Jenkinsfile

### CODE REVIEW

File: Jenkinsfile [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+          sh "gcloud config set compute/zone ${env.ZONE}"
+         }
+        }
+    }
+    // linter testing
+    stage('linter') {
+      steps {
+        container('k8s-node') {
+          // Checkout code from repository
+          checkout scm
```
Comment (techgnosis) [inline Jenkinsfile]: You don't need this line. It already checks out the code

File: Jenkinsfile [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+          sh "make create"
+        }
+      }
+    }
+    // validate infrastructure
+    stage('validate') {
+      steps {
+        container('k8s-node') {
+          script {
+            for (int i = 0; i < 3; i++) {
```
Comment (techgnosis) [inline Jenkinsfile]: Can you add a comment about why you are running validate in this loop?

File: delete.sh [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+# limitations under the License.
+
+# "---------------------------------------------------------"
+# "-                                                       -"
+# "-  Delete uninstalls Cassandra and deletes              -"
+# "-  the GKE cluster                                      -"
+# "-                                                       -"
+# "---------------------------------------------------------"
+
+# Do not set errexit as it makes partial deletes impossible
```
Comment (techgnosis) [inline delete.sh]: Could you elaborate on this comment too?

File: validate.sh [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+APP_MESSAGE="deployment \"hello-server\" successfully rolled out"
+
+# Get the kubectl credentials for the GKE cluster.
+gcloud container clusters get-credentials "$CLUSTER_NAME" --zone "$ZONE" --project "$PROJECT"
+
+# Wait for the rollout of hello-server to finish
+while true
+do
+  ROLLOUT=$(kubectl --namespace default rollout status --watch=false deployment/hello-server) &> /dev/null
```
Comment (techgnosis) [inline validate.sh]: This could loop forever if you don't get APP_MESSAGE. What if it errors instead?

File: validate.sh [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+# Grab the external IP and port of the service to confirm that hello-server deployed correctly.
+EXT_IP=""
+EXT_PORT=""
+while true
+do
+  sleep 1
+  EXT_IP="$(kubectl --namespace default get svc hello-server -ojsonpath='{.status.loadBalancer.ingress[0].ip}')"
+  EXT_PORT=$(kubectl --namespace default get service hello-server -o=jsonpath='{.spec.ports[0].port}')
+
+  if [[ $EXT_IP =~ ^[0-9]+\.[0-9]+\.[0-9]+\.[0-9]+$ ]]; then
```
Comment (techgnosis) [inline validate.sh]: Only one of these if statements has to succeed to break the loop. Do you want both? Also is EXT_PORT being tested as an IP address?

File: Jenkinsfile (hunk 1 above)
Comment (pydevops) [reply Jenkinsfile]: Good point

File: Jenkinsfile (hunk 2 above)
Comment (pydevops) [reply Jenkinsfile]: Yes

File: delete.sh (hunk 3 above)
Comment (pydevops) [reply delete.sh]: Yes I put my understanding in the comment.

File: validate.sh (hunk 4 above)
Comment (pydevops) [reply validate.sh]: Good catch. Let me do one pass validation here.

File: validate.sh (hunk 5 above)
Comment (pydevops) [reply validate.sh]: EXT_PORT is part of the URL. 
Good catch on my copy & paste error.

**Comment by pydevops** [comment]
Ok to test

**Comment by pydevops** [comment]
ok to test

**Comment by pydevops** [comment]
test this please

File: validate.sh (hunk 5 above)
Comment (pydevops) [reply validate.sh]: I added timeout with 10 minutes on the Jenkinsfile in case the 2nd loop never exits.

File: Jenkinsfile [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+spec:
+  containers:
+  - name: k8s-node
+    image: gcr.io/pso-helmsman-cicd/jenkins-k8s-node:1.0.1
+    imagePullPolicy: Always
+    command:
+    - cat
+    tty: true
+    volumeMounts:
+    # Mount the docker.sock file so we can communicate wth the local docker
```
Comment (chrislovecnm) [inline Jenkinsfile]: Typo with

File: common.sh [hunk 7]
```diff
@@ -0,0 +1,62 @@
+#!/bin/bash -e
```
Comment (chrislovecnm) [inline common.sh]: Need to use env

File: Jenkinsfile [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+  # Create a volume that contains the dev json key that was saved as a secret
+  - name: dev-key
+    secret:
+      secretName: jenkins-deploy-dev-infra
+"""
+    }
+  }
+
+// required by Terraform
+environment {
```
Comment (chrislovecnm) [inline Jenkinsfile]: Need to update this to make it look like the one Jeff is using

File: create.sh [hunk 9]
```diff
@@ -0,0 +1,40 @@
+#!/bin/bash -e
+
```
Comment (chrislovecnm) [inline create.sh]: Env please

File: delete.sh [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+# If it is set, the script would exit if any statement returns a non-true return value
+# that means it won't destory other resources if the flow continues.
+set -o nounset
+set -o pipefail
+
+ROOT="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
+# shellcheck disable=SC1090
```
Comment (chrislovecnm) [inline delete.sh]: We should not disable this

File: validate.sh [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+do
+  sleep 1
+  EXT_IP="$(kubectl --namespace default get svc hello-server -ojsonpath='{.status.loadBalancer.ingress[0].ip}')"
+  EXT_PORT=$(kubectl --namespace default get service hello-server -o=jsonpath='{.spec.ports[0].port}')
+
+  if [[ $EXT_IP =~ ^[0-9]+\.[0-9]+\.[0-9]+\.[0-9]+$ && $EXT_PORT =~ ^[0-9]+$ ]]; then
+    break
+  else
+    continue
+  fi
```
Comment (chrislovecnm) [inline validate.sh]: This is an infinite loop. Please test for like 10 min then break the loop.

Comment (pydevops) [reply validate.sh]: or set a retry count like in istio-shared/verify-functions.sh?

File: common.sh (hunk 7 above)
Comment (pydevops) [reply common.sh]: changed for all the scripts in this repo

File: delete.sh (hunk 10 above)
Comment (pydevops) [reply delete.sh]: If we don't disable it, how would we pass the shellcheck linting? any example to make it passing? 
Other repos has it too such as gke-cloud-sql-postgres and stateful?

File: create.sh (hunk 9 above)
Comment (pydevops) [reply create.sh]: changed

File: Jenkinsfile (hunk 8 above)
Comment (pydevops) [reply Jenkinsfile]: changed

File: Jenkinsfile (hunk 6 above)
Comment (pydevops) [reply Jenkinsfile]: Not sure where it is.

**Comment by pydevops** [comment]
test this please

**Comment by pydevops** [comment]
I have made changes according to the reviews @chrislovecnm @techgnosis

**Pull request merged**